myia/opt/rewrite.py

```diff
@@ -1,70 +1,472 @@
+
+        for node, key in g.higher_order_sites:
+            if not (node.is_apply(P.switch) and (key == 2 or key == 3)):
+                return None
+
+            g2_node = node.inputs[5 - key]
+            if not g2_node.is_constant_graph():
+                return None
+
+            g2 = g2_node.value
+            uses = self.manager.uses[node]
+            if not all(key2 == 0 for site, key2 in uses):
+                return None
+
+            call_sites.update({site: {g, g2} for site, _ in uses})
+
+        return call_sites
+
+    def _make_group(self, g, entries):
+        valid = True
+        ws = WorkSet([g])
+        # Set of graphs that must all be augmented with the same fvs
+        eqv = OrderedSet([g])
+        # Call sites that will have to be updated
+        call_sites = {}
+        for currg in ws:
+            eqv.add(currg)
+            new_results = self.call_sites(currg)
+            if new_results is None:
+                valid = False
+                break
+            for _, new_eqv in new_results.items():
+                ws.queue_all(new_eqv)
+            call_sites.update(new_results)
+        if valid:
+            for gg in eqv:
+                entries[gg] = NS(graph=gg, calls=call_sites, eqv=eqv)
+                # Only the first graph gets the call sites
+                call_sites = {}
+
+    def make_groups(self):
+        """Group the graphs according to their uses.
+
+        Returns {graph: entry}.
+
+        Each resulting entry contains the following fields:
+
+        * graph: The graph for this entry.
+        * eqv: A set of graphs that must be rewritten identically.
+        * calls: A {call_site: graphs} dict that maps a node to the
+            set of graphs that may be called there. That set may not
+            include the graph for this entry. Only one graph in the
+            eqv set will have a non-empty dictionary here.
+        """
+        entries = {}
+        for g in self.graphs:
+            if g in entries:
+                continue
+            self._make_group(g, entries)
+        return entries
+
+    def run(self):
+        """Run the rewriting optimization."""
+        # 1. Group the graphs and generate corresponding entries.
+        entries = self.make_groups()
+
+        # 2. Filter the entries to remove invalid ones. If an entry is invalid,
+        # all entries corresponding to graphs in its eqv are also removed,
+        # because it's all or nothing for an eqv group.
+        if not getattr(self.filter, "noop", False):
+            new_entries = dict(entries)
+            for g, entry in entries.items():
+                if g in new_entries and not self.filter(entry, entries):
+                    for gg in entry.eqv:
+                        del new_entries[gg]
+            entries = new_entries
+
+        # 3. Sort the entries, if operations must be done in a specific order
+        # if self.order_key is NotImplemented:
```

Comment:
This comment seems like leftover code.

Reply:
Yeah. It just... blended in, so I missed it :P